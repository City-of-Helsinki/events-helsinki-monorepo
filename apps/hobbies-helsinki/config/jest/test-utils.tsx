import type { ApolloCache, InMemoryCache } from '@apollo/client/cache';
import type { MockedResponse } from '@apollo/client/testing';
import type { RenderResult } from '@testing-library/react';
import { act, fireEvent, render } from '@testing-library/react';
import type { NextRouter } from 'next/router';
import Router from 'next/router';
import type { ReactElement } from 'react';
import React from 'react';
import wait from 'waait';

import { hobbiesApolloClient } from '../../src/domain/clients/hobbiesApolloClient';
import TestProviders from './TestProviders';

type CustomRender = {
  (
    ui: React.ReactElement,
    options?: {
      mocks?: MockedResponse[];
      cache?: ApolloCache<Record<string, unknown>> | InMemoryCache;
      routes?: string | string[];
    }
  ): CustomRenderResult;
};

export type CustomRenderResult = RenderResult & { router: NextRouter };

export const arrowUpKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 38, key: 'ArrowUp' });

export const arrowDownKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 40, key: 'ArrowDown' });

export const enterKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 13, key: 'Enter' });

export const escKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 27, key: 'Escape' });

export const tabKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 9, key: 'Tab' });

const customRender: CustomRender = (
  ui: ReactElement,
  { mocks = [], cache = hobbiesApolloClient.cache, routes = [] } = {}
) => {
  if (routes) {
    if (!Array.isArray(routes)) {
      routes = [routes];
    }
    routes.forEach((route) => Router.push(route));
  }

  // set locales so that the routing tests for default locale worked
  Router.locales = ['fi', 'en', 'sv'];

  const renderResult = render(ui, {
    wrapper: ({ children }) => (
      <TestProviders mocks={mocks} router={Router} cache={cache}>
        {children}
      </TestProviders>
    ),
  });
  return {
    ...renderResult,
    router: Router,
  };
};

// eslint-disable-next-line testing-library/no-unnecessary-act
const actWait = (amount?: number): Promise<void> => act(() => wait(amount));

// eslint-disable-next-line import/export
export { actWait, customRender as render };

// re-export everything
// eslint-disable-next-line import/export
export * from '@testing-library/react';
export { render as defaultRender } from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
