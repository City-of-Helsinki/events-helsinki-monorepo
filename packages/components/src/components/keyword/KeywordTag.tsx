import classNames from 'classnames';
import type { FunctionComponent } from 'react';
import { useState } from 'react';
import { TagComponent as Tag } from 'react-helsinki-headless-cms';

import styles from './keyword.module.scss';

interface Props {
  blackOnMobile?: boolean;
  whiteOnly?: boolean;
  featured?: boolean;
  transparent?: boolean;
  hideOnMobile?: boolean;
  keyword: string;
  onClick?: () => void;
  className?: string;
}

const KeywordTag: FunctionComponent<Props> = ({
  blackOnMobile,
  whiteOnly,
  hideOnMobile,
  transparent,
  keyword,
  className,
  onClick,
  ...rest
}) => {
  const [isSelected, setIsSelected] = useState(false);

  const handleClick = (): void => {
    setIsSelected(true);
    onClick && onClick();
  };

  return (
    <Tag
      whiteOnly={whiteOnly}
      selected={isSelected}
      className={classNames(
        styles.keyword,
        blackOnMobile && styles.blackOnMobile,
        hideOnMobile && styles.hideOnMobile,
        transparent && styles.transparent,
        className
      )}
      onClick={handleClick}
      {...rest}
    >
      {keyword}
    </Tag>
  );
};

export default KeywordTag;
