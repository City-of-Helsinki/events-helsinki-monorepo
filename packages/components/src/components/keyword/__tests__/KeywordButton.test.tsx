import userEvent from '@testing-library/user-event';
import React from 'react';
import { render, screen } from '@/test-utils';
import KeywordTag from '../KeywordTag';

const keyword = 'test keyword';

it('matches snapshot', () => {
  const { container } = render(
    <KeywordTag keyword={keyword} onClick={vi.fn()} />
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('calls onClick callback when clicking', async () => {
  const onClickMock = vi.fn();
  render(<KeywordTag keyword={keyword} onClick={onClickMock} />);

  expect(screen.getByText(keyword)).toBeInTheDocument();

  await userEvent.click(screen.getByRole('link'));

  expect(onClickMock).toHaveBeenCalled();
});
