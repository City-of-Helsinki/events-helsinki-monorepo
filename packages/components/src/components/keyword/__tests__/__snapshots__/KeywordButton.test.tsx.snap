// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="Tag-module_tag__3bo5S tag_hds-tag__3MKf3 custom-theme-1 tag-module_tag__1QCQs tag-module_withHover__3daxF keyword"
  id="tag:r0:"
  role="link"
  tabindex="0"
>
  <span
    class="Tag-module_label__zCgFr undefined"
    id="tag:r0:-label"
  >
    <span>
      test keyword
    </span>
  </span>
</div>
`;
